### ISSUE TITLE
NaN loss and accuracy when using MirroredStrategy on multiple GPUs (ROCm)

### ISSUE THREAD

**Issue by Laggger164** [tensorflow/tensorflow#41657]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): 
No, the code is from this example: https://www.tensorflow.org/tutorials/distribute/keras
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): 
Linux Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: 
No
- TensorFlow installed from (source or binary): 
ROCm 3.5 
- TensorFlow version (use command below): 
v2.2.0-30-g34c3143 2.2.0
- Python version: 
3.6.9
- Bazel version (if compiling from source): 
No
- GCC/Compiler version (if compiling from source): 
No
- CUDA/cuDNN version: 
ROCm
- GPU model and memory: 
2x AMD VEGA 10 XT (Vega 64) 8GB 

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**

Mirrored strategy causes the following loss and accuracy values:

Epoch 12/12
468/469 [============================>.] - ETA: 0s - loss: nan - accuracy: nan
469/469 [==============================] - 12s 26ms/step - loss: nan - accuracy: nan - lr: 1.0000e-05
79/79 [==============================] - 1s 12ms/step - loss: nan - accuracy: nan
Eval loss: nan, Eval Accuracy: nan

This seems to happen on Nvidia cards too:
https://github.com/tensorflow/tensorflow/issues/36224

I have tried several other scripts and all do the same.

**Describe the expected behavior**

Running on a single card produces an expected output of loss and accuracy.
I would like that to be the case in multiple GPUs too.

Not sure whether it's just the loss and accuracy calculations screwed up, as the training goes about twice as fast on 2 GPUs.

**This code seems to work flawlessly on both GPUs:**

[mnistmultigpu.txt](https://github.com/tensorflow/tensorflow/files/4965502/mnistmultigpu.txt)
However, it only seems to utilize 1 GPU, which is strange because it did utilize both before...

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

Unfortunately I cannot guarantee you will be able to reproduce this as this is tied to having multiple GPUs (Nvidia or not)

<details><summary> The code I used </summary>
# Import TensorFlow and TensorFlow Datasets

import tensorflow_datasets as tfds
import tensorflow as tf
tfds.disable_progress_bar()

import os

datasets, info = tfds.load(name='mnist', with_info=True, as_supervised=True)

mnist_train, mnist_test = datasets['train'], datasets['test']

strategy = tf.distribute.MirroredStrategy()
print('Number of devices: {}'.format(strategy.num_replicas_in_sync))

# You can also do info.splits.total_num_examples to get the total
# number of examples in the dataset.

num_train_examples = info.splits['train'].num_examples
num_test_examples = info.splits['test'].num_examples

BUFFER_SIZE = 10000

BATCH_SIZE_PER_REPLICA = 64
BATCH_SIZE = BATCH_SIZE_PER_REPLICA * strategy.num_replicas_in_sync

def scale(image, label):
  image = tf.cast(image, tf.float32)
  image /= 255

  return image, label

train_dataset = mnist_train.map(scale).cache().shuffle(BUFFER_SIZE).batch(BATCH_SIZE)
eval_dataset = mnist_test.map(scale).batch(BATCH_SIZE)

with strategy.scope():
  model = tf.keras.Sequential([
      tf.keras.layers.Conv2D(32, 3, activation='relu', input_shape=(28, 28, 1)),
      tf.keras.layers.MaxPooling2D(),
      tf.keras.layers.Flatten(),
      tf.keras.layers.Dense(64, activation='relu'),
      tf.keras.layers.Dense(10)
  ])

  model.compile(loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
                optimizer=tf.keras.optimizers.Adam(),
                metrics=['accuracy'])

# Define the checkpoint directory to store the checkpoints

checkpoint_dir = './training_checkpoints'
# Name of the checkpoint files
checkpoint_prefix = os.path.join(checkpoint_dir, "ckpt_{epoch}")

# Function for decaying the learning rate.
# You can define any decay function you need.
def decay(epoch):
  if epoch < 3:
    return 1e-3
  elif epoch >= 3 and epoch < 7:
    return 1e-4
  else:
    return 1e-5

# Callback for printing the LR at the end of each epoch.
class PrintLR(tf.keras.callbacks.Callback):
  def on_epoch_end(self, epoch, logs=None):
    print('\nLearning rate for epoch {} is {}'.format(epoch + 1,
                                                      model.optimizer.lr.numpy()))

callbacks = [
    tf.keras.callbacks.TensorBoard(log_dir='./logs'),
    tf.keras.callbacks.ModelCheckpoint(filepath=checkpoint_prefix,
                                       save_weights_only=True),
    tf.keras.callbacks.LearningRateScheduler(decay),
    PrintLR()
]

model.fit(train_dataset, epochs=12, callbacks=callbacks)

model.load_weights(tf.train.latest_checkpoint(checkpoint_dir))

eval_loss, eval_acc = model.evaluate(eval_dataset)

print('Eval loss: {}, Eval Accuracy: {}'.format(eval_loss, eval_acc))

</details>

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

[runlog.txt](https://github.com/tensorflow/tensorflow/files/4965388/runlog.txt)

My code in a text file:
[mirroredstrategy.txt](https://github.com/tensorflow/tensorflow/files/4965443/mirroredstrategy.txt)

#36224

Any and all help would be appreciated, thank you!

**Comment by Laggger164** [comment]
> **This code seems to work flawlessly on both GPUs:**
> 
> [mnistmultigpu.txt](https://github.com/tensorflow/tensorflow/files/4965502/mnistmultigpu.txt)
> However, it only seems to utilize 1 GPU, which is strange because it did utilize both before...

#30321
It seems that it may have worked properly when I experimented with an older version of TensorFlow.

So basically even that one doesn't work.

**Comment by nikitamaia** [comment]
Hi @Laggger164, just to make sure I'm clear on this, you are running the basic Distributed training with Keras MNIST example. When you have just one GPU with MirroredStrategy, everything works. However, when you add a second GPU you are seeing NaN loss and accuracy. Is that a correct summary?

I'm not able to reproduce this error. So I'm wondering if maybe there's an issue with one of your GPUs? If you have more than one GPU in your system, the GPU with the lowest ID will be selected by default. Can you try and train with one GPU but enforce that it's GPU:1 instead of GPU:0? You can do that with tf.config.set_visible_devices to specify which devices are visible to the runtime.

**Comment by Laggger164** [comment]
> Hi @Laggger164, just to make sure I'm clear on this, you are running the basic Distributed training with Keras MNIST example. When you have just one GPU with MirroredStrategy, everything works. However, when you add a second GPU you are seeing NaN loss and accuracy. Is that a correct summary?
Pretty much, yes. Thing is it goes about twice as fast with 2 GPUs so it is possible that it works correctly, just the loss and accuracy metrics are calculated incorrectly.

> I'm not able to reproduce this error. So I'm wondering if maybe there's an issue with one of your GPUs? If you have more than one GPU in your system, the GPU with the lowest ID will be selected by default. Can you try and train with one GPU but enforce that it's GPU:1 instead of GPU:0? You can do that with tf.config.set_visible_devices to specify which devices are visible to the runtime.

I tried with both GPUs, both seem to work identically well.
Here are the logs:

[mirroredstrategysinglegpu1.txt](https://github.com/tensorflow/tensorflow/files/4970727/mirroredstrategysinglegpu1.txt)
[mirroredstrategysinglegpu2.txt](https://github.com/tensorflow/tensorflow/files/4970728/mirroredstrategysinglegpu2.txt)

And I tried modifying the code to remove the mirrored strategy:

[strategynomirror.txt](https://github.com/tensorflow/tensorflow/files/4970733/strategynomirror.txt)
and the logs from that:
[strategysinglegpu1.txt](https://github.com/tensorflow/tensorflow/files/4970735/strategysinglegpu1.txt)
[strategysinglegpu2.txt](https://github.com/tensorflow/tensorflow/files/4970736/strategysinglegpu2.txt)

Seemed to be just a little faster (9 seconds instead of 11) with just one GPU than the one with mirrored strategy.

I used HID_VISIBLE_DEVICES=0 and 1, verifying the usage with rocm-smi on watch.

So I believe there is a problem in how either the code or the system is handling mirrored strategy.
I could try reinstalling the OS with a different version of everything, then upgrading to the newer ones, however that seems like a shot in the dark. I have no idea what I am doing basically...

Also, you mentioned you aren't able to reproduce this, meaning you don't have the hardware or system, or did you try and it is working on your end?

**Comment by Laggger164** [comment]
It seems that I managed to solve this by reinstalling everything.

One thing that I _did_ change was that I didn't install the AMDGPU-PRO drivers before installing ROCm.

Which I later found out was completely useless as ROCm is another type of driver, so it is possible I screwed something up that way.

I still have a problem with the mirrorStrategy as it is only utilizing multiple GPUs to exactly half their power when using 2 and exactly a quarter when using 4.

Pretty sure this could be caused by some setting in Tensorflow that I don't yet know of.
If you do know then I would be thankful if you could tell me.


Not sure if we should close this issue at this point and make another one. You decide 👍

**Comment by nikitamaia** [comment]
Glad to hear that reinstalling worked!
As for the performance issue, can you explain a bit more about what you mean by using half their power? Are you seeing a slowdown with the more GPUs you add?

**Comment by Laggger164** [comment]
> Glad to hear that reinstalling worked!
> As for the performance issue, can you explain a bit more about what you mean by using half their power? Are you seeing a slowdown with the more GPUs you add?

Yes, that's exactly what happens.

However it did work better with both of the cards connected to full PCIe 16x slots... which improved the training time by about 2 seconds from 17 seconds.
As opposed to using PCIe 1x risers, which just slow it down.

As for the power I mean GPU utilization and GPU power draw, metric taken from rocm-smi on watch.
i.e. when using a single GPU both are at 100%, but when using 2, both are at around 50%. VRAM being always full.

It's basically telling me that it _could work_ but doesn't right now.

I also noticed the CPU utilization was almost always at 100% with 4 cards running, while being at around 70% with 2 cards.
I am wondering whether there is a bandwidth issue, the CPU and motherboard are pretty cheap so I will try a much stronger setup at some point.

But if you have any ideas, feel free to tell me!

**Comment by nikitamaia** [comment]
If you're still facing this utilization problem, can you file a new issue? Any Tensorboard Profiler info you can provide as well will be helpful to debug. There are guides [here](https://www.tensorflow.org/tensorboard/tensorboard_profiling_keras) and [here](https://www.tensorflow.org/guide/profiler) on how to use the tool.

**Comment by nikitamaia** [comment]
Closing this issue now since original problem was solved. If performance issues persist, please file a new issue and I'd be happy to help debug!

**Comment by mepster** [comment]
This issue still exists in tf / keras 2.8.0. When you are running on multiple GPUs, sometimes an empty batch is created. Depending on your loss function, this can create a divide- by-zero error. 

There is a workaround described here:
https://stackoverflow.com/questions/54283937/training-on-multiple-gpus-causes-nan-validation-errors-in-keras

Quoting from that link: "As far as I can (heuristically) tell, when doing distributed training/evaluation, the number of elements in the dataset must be evenly divisible by the batch size and number of GPUs. That is, nelements / ngpus / batch_size == 0. If that is not the case, then empty batches will be passed to the loss function, which, depending on the loss function, may inject NaN losses into the aggregator."

That guy's explanation matches what I see, and his workaround (checking if the batch size is zero and in that case avoiding a divide-by-zero) works for me. But it should really be fixed in tf - empty batches should not be created.

Like OP I am using tf.distribute.MirroredStrategy and "with strategy.scope()".

I believe this is the same issue: https://github.com/tensorflow/tensorflow/issues/36224
